### PR TITLE
ProtoBuf metadata

### PR DESCRIPTION
**Pull request by bachdavi**
This PR adds support for our new ProtoBuf metadata. 

## Review Guidelines
- I bumped the version to `0.2.0` and added missing `CHANGELOG` entries
- The proto files are located in `proto/` and are copied from our RAI repository
- I've used our new, internally developed `ProtocolBuffers.jl` to generate the Julia structs, these can be found here: `src/gen/`
- We now **only** use the new protobuf metadata by setting the respective header
- The changes to the handling is straight-forward; we decode the multi-part in the fast-path as ProtoBuf and the body in `get_transaction_metadata`
- I added `metadata.jl` that adds methods to handle the new metadata, in this PR I only added a `show_*` method that prints a `RelationId` in a nice way, e.g. `/ConstantType(Symbol, :output)/Int64` or `/ConstantType(Symbol, :output)/ValueType(ConstantType(String, :rel), ConstantType(String, :base), ConstantType(String, :Hash), Int128)` or when specializing on a `ValueType` : `/ConstantType(Symbol, :output)/ConstantType(ValueType(ConstantType(String, :rel), ConstantType(String, :base), ConstantType(String, :Hash), Int128), 0xc0793bb4b6967d16193e6dff360f6d63)`

### CODE REVIEW

File: CHANGELOG.md
Comment (NHDaly) [inline CHANGELOG.md:24]: Wow, thank you for updating the changelog! ❤️

File: Project.toml
Comment (NHDaly) [inline Project.toml:14]: LEAVING A REMINDER HERE TO UPDATE TO `ProtoBuf` ONCE IT'S MERGED THERE

File: src/RAI.jl
Comment (NHDaly) [inline src/RAI.jl:98]: I guess shouldn't this be including `Relationalai_PB.jl` instead?
I guess maybe you could do this:
```suggestion
include("gen/relationalai/Relationalai_PB.jl")
using .Relationalai_PB: Protocol_PB
```

File: src/metadata.jl
Comment (NHDaly) [inline src/metadata.jl:72]: This is very nifty! 😁 👍 

Personally, i think that the `/`s are a bit out-dated, and we should just leave them behind entirely. They don't correspond to anything in Rel, they're a bit distracting, and they don't stand out compared to the whitespace in an example like this:
`/ConstantType(Symbol, :output)/Int64`, where the whitespace is a more obvious separator than the `/`!

I would suggest that we just format this as a normal-looking julia array, like this:
```julia
[ConstantType(Symbol, :output), Int64]
```
Another nice thing about this format is that we could make it an actually valid copy/pastable representation, I think?

File: src/api.jl
Comment (NHDaly) [inline src/api.jl:483]: 👍 cool, thanks, makes sense! 👍

File: src/results.jl
Comment (NHDaly) [inline src/results.jl:90]:
```suggestion
        if idx !== lastindex(rsp.metadata.relations)
```
After reading Yuri's blog post (https://yuri.is/not-julia/), I think the ask from the julia community is to try to always compare iterators using the "index" API, rather than directly checking against an array's length. The idea is that this code will be resilient to using slices and other weird arrays that aren't indexed `1:N`. I'm not sure if it really matters for this code, which is not generic, but it's maybe nicer to read anyway, and seems like a good habit to get into? 😊

File: src/metadata.jl
Comment (NHDaly) [reply src/metadata.jl:72]: Although it's maybe then weird that the `ConstantType(Symbol, :output)` struct doesn't actually exist. ... I was going to suggest what if we _made_ it exist, and we could maybe print it like:
```julia
[ConstantType{Symbol}(:output), Int64]
```
But i think it's maybe not useful to have both the proto structs _and_ separate julia structs that we convert the proto structs into? 🤔 i'm not sure; i'd need to think more about it.

Anyway, maybe what you have here is already good for now, and we can think more about it as we go on.

Maybe we could just drop the whitespace in there? Or maybe _add_ whitespace between the `/`s?:
- `/ ConstantType(Symbol, :output) / Int64`
- `/ConstantType(Symbol,:output)/Int64`

File: src/metadata.jl
Comment (NHDaly) [inline src/metadata.jl:23]: Is there any reason to map to these julia types, rather than just mapping to the types as strings? e.g. 
```julia
    Protocol_PB.PrimitiveType.INT_64 => "Int64",
```
?
Are the types used anywhere other than for printing? I always get nervous when I see types used as values, because it tends to cause both dynamic and compilation heavy code

Comment (NHDaly) [reply src/metadata.jl:23]: Ah, i see that it's used for `from_proto()`, for constant values, below.. 🤔 Yeah maybe how you have it is right..

I do wonder if it could just be a big switch-statement though. Might be interesting to profile the two cases some day, for schema-heavy queries?

**Review by NHDaly — APPROVED** [review]
Awesome, thanks @bachdavi! 🎉 

As a reminder, I think we should _hold off_ on merging until ProtocolBuffers.jl is released.
In the meantime, we can leave this PR sitting, waiting and ready, or we can try to switch back to ProtoBuf if we find out that it will be a long time.

Thanks again! :)

**Review by NHDaly — CHANGES_REQUESTED** [review]
Blocking the PR until the ProtoBuf situation is resolved

File: src/metadata.jl
Comment (NHDaly) [inline src/metadata.jl:93]: I think we should special-case `ConstantType(String,x)` and `ConstantType(Symbol,x)` to display as just `:x`. This is the julia equivalent, and it's how it's written in Rel, and it's much nicer looking. 😁 

We can also maybe write other constant values as `#(x)`, but i'm not sure. we should talk about it!

Comment (NHDaly) [reply src/metadata.jl:93]: One of the downsides of `#(x)` is that you lose the type info... So maybe it should be: `#(Int(5))`, `#(Int8(2))`, etc?

File: test/api.jl
Comment (NHDaly) [inline test/api.jl:171]: is this still true? i thought that with the new Protos package we _can_ test for equality now?

File: src/metadata.jl
Comment (NHDaly) [reply src/metadata.jl:93]: Or maybe just `Int8(5)` directly in the metadata. That's how it's printed in the console now:
Screen Shot 2022-08-24 at 6 01 18 PM

**Review by NHDaly — APPROVED** [review]
Okay, now that this has been updated to ProtoBuf 1.0 (thanks @Drvi!) I think this is good to merge! Sorry for my delay. 

I will merge it now, and then we can address all of my open questions in follow ups! 😁

**Comment by NHDaly** [comment]
Oh except JK, the tests are failing 😅

**Comment by NHDaly** [comment]
Okay fixed it, it was a merge conflict with the support for printing `true` relations (#15).

File: test/api.jl
Comment (Drvi) [reply test/api.jl:171]: Sadly, the default `==` comparison won't work if the proto contains e.g. arrays:
```julia
julia> struct MyStruct
         a::Int
         b::Vector{Int}
       end

julia> MyStruct(1, [2,3]) == MyStruct(1, [2,3])
false
```
If we want to, we could tweak the new codegen to optionally prepend everything with `@auto_hash_equals` or even generate the `==` and `hash` methods ourselves.

**Comment by Drvi** [comment]
I'll have a look at the failing test cases, seems like the printing is not working right as now some of the protos structs are parametrized.

**Comment by Drvi** [comment]
Looking at it now, are the failing tests really failing? We're getting:
```
/ConstantType(String, "output")/ConstantType(String, "a")/Int64
 (1,)

/ConstantType(String, "output")/ConstantType(String, "b")/Int64/Int64
 (2, 3)
 (4, 5)
```
while the expected output was:
```
/:output/:a/Int64
 (1,)

/:output/:b/Int64/Int64
 (2, 3)
 (4, 5)
```

**Comment by NHDaly** [comment]
Ah, makes sense! I guess this is a "soft merge conflict," because those tests didn't exist when @bachdavi first wrote this PR.

Now that it's already the friday of the last day of david's vacation, maybe we'll wait for his return. I _think_ my preference might be to change the printing to be more like what was there before. Or something new entirely.

But i'm also supportive of just fixing the failing tests and then iterating in subsequent PRs if that's the direction we want to go. 👍

**Comment by NHDaly** [comment]
Thanks for picking that out, @Drvi!

File: src/metadata.jl
Comment (Drvi) [reply src/metadata.jl]: > I do wonder if it could just be a big switch-statement though.

Sounds like a good idea to me 👍

**Comment by bachdavi** [comment]
I'm picking this up again :) 
Thank you @NHDaly and @Drvi for working on the ProtoBuf.jl integration!
I'm not sure about the way we want to print the metadata for `show_results`, I will experiment with a few things and then see what fits best.

**Comment by NHDaly** [comment]
Thanks @bachdavi! I think one reasonable approach would be just merging now, and then we can always come back to address the output style in subsequent PRs.

I'm sorry we didn't take that approach while you were out to get this over the finish line!

File: src/results.jl
Comment (bachdavi) [reply src/results.jl:91]: Love it :)

File: src/metadata.jl
Comment (bachdavi) [reply src/metadata.jl]: I leave this as is for now, but add a `TODO` for the future. We could change `_from_proto` to use a big switch statement indeed :)

File: src/metadata.jl
Comment (bachdavi) [reply src/metadata.jl:73]: See https://relationalai.slack.com/archives/C02C9AV2LL9/p1661963216091409

File: src/metadata.jl
Comment (bachdavi) [reply src/metadata.jl]: See https://relationalai.slack.com/archives/C02C9AV2LL9/p1661963216091409

**Review by NHDaly — APPROVED** [review]

**Pull request merged**
